Add helpers for Nov 2019 protocol upgrade MTP based code activation

sickpig opened the pull request:
based on top #1900

~~TODO add unit test~~ (done)

sickpig committed:
Add helpers to use in Nov 2019 protocol upgrade MTP based code activation

sickpig committed:
Clear mempool and txs queue in case we get a reorg across Nov 2019 block activation

dagurval commented on src/chainparams.cpp:
```diff
@@ -20,7 +20,7 @@
 
 // FIXME This need to be update every new fork
-uint64_t nMiningForkTime = 1557921600;
+uint64_t nMiningForkTime = 1573819200;
```
Nit: How about a `const uint64_t GRAVITON_ACTIVATION_TIME = 1573819200`, rather than repeating this number 4 times in this file?

sickpig replied:
makes sense. having a constant would be better, even thou I'd rather chose another name for it, "gravitation" it is ABC specific.

ptschip replied:
const seems fine but what's wrong with nMiningForkTime...IMO seems more intuitive than some other name.

sickpig replied:
@ptschip `nMiningForkTime` is used in the definition of the `miningForkTime` so it can't be `const`. Having said that in commit  
9bc2e24 I used this var to initliaze all chain params related to activation time

gandrewstone replied:
Use NOV2019_ACTIVATION_TIME.  Using a code name here makes no sense, and may be confusing in the future (if we don't remove the constant post upgrade) since it will be hard to remember what code name goes to what.

dagurval commented on src/validation/forks.cpp:
```diff
@@ -95,3 +95,21 @@ bool IsNov2018Activated(const Consensus::Params &consensusparams, const CBlockIn
 }
+
+bool IsNov2019Enabled(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
+{
```
`pindexTip`? This looks to be parent of tip. So `prevTip`? `parentOfTip`?

sickpig replied:
`IsNov2019Enabled` is wrapper around `IsforkActiveOnNextBlock` which return `true` if the fork will be enabled at the next block in respect to the current tip (see `src/chain.cpp`).

So if we use `c` as a label for the activation block, this function will return true for every block `x` with height that belongs to the interval `[h(c-1),+inf]`. (`h()` return the height of a block).

You could use that helper with any block, this is not meant to be used with current tip or the parent of the current tip. 

E.g. in `DisconnectTip()` we used twice, once with the block that is going to be disconnected and the with its parent. 

```c++
if (IsNov2019Enabled(consensusParams, pindexDelete) && 
       !IsNov2019Enabled(consensusParams, pindexDelete->pprev))
```

So that the condition above will return true only if we are disconnecting the block which activated the fork, i.e. we have a reorg across fork activation block.

dagurval commented on src/validation/validation.cpp:
```diff
@@ -2915,6 +2915,15 @@ bool DisconnectTip(CValidationState &state, const Consensus::Params &consensusPa
+    // not previously avaiable features (e.g. Schnorr enabled OP_CHECKMULTISIG)
+    if (IsNov2019Enabled(consensusParams, pindexDelete) && !IsNov2019Enabled(consensusParams, pindexDelete->pprev))
+    {
+        mempool.clear();
```
Is there a mempool admittance queue that also needs to cleared? ping @gandrewstone

sickpig replied:
good question? I think it all depends on the fact if the txs in the queue are supposed to be already validate or not. Agree that @gandrewstone is the right person to ask to.

ptschip replied:
yes there's a commitQ.   You can do this to clear it.

+boost::unique_lock<boost::mutex> lock(csCommitQ);
+txCommitQ->clear();

ptschip replied:
on the other hand if 1902 gets merged then there's no need to clear the mempool here.

sickpig replied:
@ptschip agreed. Will add the bit about `txCommitQ()` and will remove it later on  in case #1902 will get merged before this one.

sickpig replied:
done

dagurval committed:
Add unit tests for Nov 2019 activation code (IsNov2019Enabled)

Grant authorship to @dagurval cause I just basically took this
code from ABC repo and adapted to work with BU, and because
the code has been authored by @dagurval originally.

sickpig committed:
Initialize nov2019ActivationTime and nMiningForkTime using a const rather than a literal

sickpig committed:
Update tweak label use to set Nov '19 mining fork time

dagurval commented on src/chainparams.cpp:
```diff
@@ -137,7 +137,7 @@ class CMainParams : public CChainParams
         consensus.may2019Height = 582679;
         // Nov, 15 2019 12:00:00 UTC fork activation time
-        consensus.nov2019ActivationTime = 1573819200;
+        consensus.nov2019ActivationTime = nMiningForkTime;
```
This is not an improvement IMO. Before, this was a constant. Now it's determined by runtime. I suggest making a constant for `1573819200` (as suggested before) or if not, leave it as it was.

sickpig replied:
Agreed. 

Fixed now (hope so)

```c++
-// FIXME This need to be update every new fork
-uint64_t nMiningForkTime = 1573819200;
+// Next protocol upgrade will be activated once MTP >= Nov 15 12:00:00 UTC 2019
+const uint64_t NOV2019_ACTIVATION_TIME = 1573819200;
+uint64_t nMiningForkTime = NOV2019_ACTIVATION_TIME;
 
 CBlock CreateGenesisBlock(CScript prefix,
     const std::string &comment,
@@ -137,7 +138,7 @@ public:
         // May, 15 2019 hard fork
         consensus.may2019Height = 582679;
         // Nov, 15 2019 12:00:00 UTC fork activation time
-        consensus.nov2019ActivationTime = 1573819200;
+        consensus.nov2019ActivationTime = NOV2019_ACTIVATION_TIME;
```

gandrewstone replied:
this allows for an emergency abort -- if some last minute issue comes up, people can configure it to sometime in the future and not fork.

sickpig committed:
Fix a bash syntax error in .travis.yml
